src/gluonnlp/data/data_preprocessing_transform.py

```diff
@@ -0,0 +1,421 @@
+            tok_to_orig_index += [i] * len(sub_tokens)
+            all_doc_tokens += sub_tokens
+        # tokenize the paragraph text and align the start/end position
+        # to the tokenized sequence position
+        tok_start_position = None
+        tok_end_position = None
+        if self.is_training and example.is_impossible:
+            tok_start_position = -1
+            tok_end_position = -1
+        if self.is_training and not example.is_impossible:
+            tok_start_position = orig_to_tok_index[example.start_position]
+            if example.end_position < len(example.doc_tokens) - 1:
+                tok_end_position = orig_to_tok_index[example.end_position +
+                                                     1] - 1
+            else:
+                tok_end_position = len(all_doc_tokens) - 1
+            (tok_start_position, tok_end_position) = improve_answer_span(
+                all_doc_tokens, tok_start_position, tok_end_position,
+                self._tokenizer, example.orig_answer_text)
+        return tok_start_position, tok_end_position, all_doc_tokens, query_tokens
+
+
+class DocSpanTransform:
+    """
+    We can have documents that are longer than the maximum sequence length.
+    To deal with this we do a sliding window approach, where we take chunks
+    of the up to our max length with a stride of `doc_stride`.
+    """
+    def __init__(self, doc_stride, max_seq_length=None):
+        self._doc_stride = doc_stride
+        self._max_seq_length = max_seq_length
+
+    def __call__(self,
+                 all_doc_tokens,
+                 max_tokens_for_doc=None,
+                 query_tokens_length=None):
+        _DocSpan = collections.namedtuple(  # pylint: disable=invalid-name
+            'DocSpan', ['start', 'length'])
+        assert max_tokens_for_doc or (self._max_seq_length
+                                      and query_tokens_length)
+        doc_spans = []
+        start_offset = 0
+        if not max_tokens_for_doc:
+            max_tokens_for_doc = self._max_seq_length - query_tokens_length - 3
+        while start_offset < len(all_doc_tokens):
+            length = len(all_doc_tokens) - start_offset
+            if length > max_tokens_for_doc:
+                length = max_tokens_for_doc
+            doc_spans.append(_DocSpan(start=start_offset, length=length))
+            if start_offset + length == len(all_doc_tokens):
+                break
+            start_offset += min(length, self._doc_stride)
+        return doc_spans
+
+
+class SimpleQAPreparation:
+    """
+    Give the tokenized query text and doc spans, convert the data to BERT-style model input.
+    Note that this class does not check if max span.
+    """
+    def __init__(self, cls_token, sep_token, vocab, is_training):
+        self.insert = ConcatSeqTransform()
+        self.is_training = is_training
+        self._cls_token = cls_token
+        self._sep_token = sep_token
+        self._vocab = vocab
```

Comment:
Why do the mapping as part of this class? Each class here should do one thing and do it well. If it does many unrelated things, it becomes hard to understand and to reuse.